src/data.c

```diff
@@ -0,0 +1,976 @@
+
+    switch (type) {
+        case ASTARTE_MAPPING_TYPE_BINARYBLOB:
+            ESP_LOGD(TAG, "Deserializing binary blob data.");
+            ares = deserialize_binaryblob(bson_elem, data);
+            break;
+        case ASTARTE_MAPPING_TYPE_BOOLEAN:
+            ESP_LOGD(TAG, "Deserializing boolean data.");
+            bool bool_tmp = new_ast_bson_deserializer_element_to_bool(bson_elem);
+            *data = astarte_data_from_boolean(bool_tmp);
+            break;
+        case ASTARTE_MAPPING_TYPE_DATETIME:
+            ESP_LOGD(TAG, "Deserializing datetime data.");
+            int64_t datetime_tmp = new_ast_bson_deserializer_element_to_datetime(bson_elem);
+            *data = astarte_data_from_datetime(datetime_tmp);
+            break;
+        case ASTARTE_MAPPING_TYPE_DOUBLE:
+            ESP_LOGD(TAG, "Deserializing double data.");
+            double double_tmp = new_ast_bson_deserializer_element_to_double(bson_elem);
+            *data = astarte_data_from_double(double_tmp);
+            break;
+        case ASTARTE_MAPPING_TYPE_INTEGER:
+            ESP_LOGD(TAG, "Deserializing integer data.");
+            int32_t int32_tmp = new_ast_bson_deserializer_element_to_int32(bson_elem);
+            *data = astarte_data_from_integer(int32_tmp);
+            break;
+        case ASTARTE_MAPPING_TYPE_LONGINTEGER:
+            ESP_LOGD(TAG, "Deserializing long integer data.");
+            int64_t int64_tmp = 0U;
+            if (bson_elem.type == ASTARTE_BSON_TYPE_INT32) {
+                int64_tmp = (int64_t) new_ast_bson_deserializer_element_to_int32(bson_elem);
+            } else {
+                int64_tmp = new_ast_bson_deserializer_element_to_int64(bson_elem);
+            }
+            *data = astarte_data_from_longinteger(int64_tmp);
+            break;
+        case ASTARTE_MAPPING_TYPE_STRING:
+            ESP_LOGD(TAG, "Deserializing string data.");
+            ares = deserialize_string(bson_elem, data);
+            break;
+        default:
+            ESP_LOGE(TAG, "Unsupported mapping type.");
+            ares = ASTARTE_RESULT_INTERNAL_ERROR;
+            break;
+    }
+    return ares;
+}
+
+static astarte_result_t deserialize_binaryblob(
+    new_ast_bson_element_t bson_elem, astarte_data_t *data)
+{
+    astarte_result_t ares = ASTARTE_RESULT_OK;
+    uint8_t *dyn_deserialized = NULL;
+
+    uint32_t deserialized_len = 0;
+    const uint8_t *deserialized
+        = new_ast_bson_deserializer_element_to_binary(bson_elem, &deserialized_len);
+
+    dyn_deserialized = calloc(deserialized_len, sizeof(uint8_t));
+    if (!dyn_deserialized) {
+        ESP_LOGE(TAG, "Out of memory %s: %d", __FILE__, __LINE__);
+        ares = ASTARTE_RESULT_OUT_OF_MEMORY;
+        goto failure;
+    }
+
+    memcpy(dyn_deserialized, deserialized, deserialized_len);
+    *data = astarte_data_from_binaryblob((void *) dyn_deserialized, deserialized_len);
+    return ares;
+
+failure:
+    free(dyn_deserialized);
+    return ares;
+}
```

Comment:
What do you think about return directly here? It seems you don't have other case where the function can fail, so you can skip the goto e return directly the error.
```suggestion
        return ASTARTE_RESULT_OUT_OF_MEMORY;
    }

    memcpy(dyn_deserialized, deserialized, deserialized_len);
    *data = astarte_data_from_binaryblob((void *) dyn_deserialized, deserialized_len);
    return ares;   
```